pkg/registry/populator_test.go

```diff
@@ -1534,3 +1534,425 @@ func TestSemverPackageManifest(t *testing.T) {
 		})
 	}
 }
+
+func TestSubstitutesFor(t *testing.T) {
+	type args struct {
+		bundles  []string
+		rebuilds []string
+	}
+	type expected struct {
+		bundles        []string
+		channels       []string
+		channelHeads   map[string]string
+		defaultChannel string
+	}
+	tests := []struct {
+		description string
+		args        args
+		expected    expected
+	}{
+		{
+			description: "SubstitutesFor/FirstSubstitutionReplaces",
+			args: args{
+				bundles: []string{
+					"prometheus.0.22.2",
+					"prometheus.0.14.0",
+					"prometheus.0.15.0",
+				},
+				rebuilds: []string{
+					"prometheus.0.15.0.substitutesfor",
+				},
+			},
+			expected: expected{
+				bundles: []string{
+					"quay.io/test/prometheus.0.22.2",
+					"quay.io/test/prometheus.0.14.0",
+					"quay.io/test/prometheus.0.15.0.substitutesfor",
+					"quay.io/test/prometheus.0.15.0.substitutesfor",
+					"quay.io/test/prometheus.0.15.0",
+					"quay.io/test/prometheus.0.15.0",
+					"quay.io/test/prometheus.0.14.0",
+				},
+				channels: []string{
+					"stable",
+					"preview",
+					"preview",
+					"stable",
+					"stable",
+					"preview",
+					"preview",
```

Comment:
I think the expected graph would be a lot easier to understand if we organized replacement chains by channel; e.g.

```go
channels: map[string][]Replaces{ // channel name/replacement tuples in channel
    "stable": [{ next: "prometheusoperator.0.15.0+1-freshmaker-rebuild", prev: "..."}, ...],
    "preview": [{ next: "prometheusoperator.0.22.2", prev: "..."}, ...],
}
```